src/services/Functional/User.js

```diff
@@ -1,51 +1,41 @@
-      if (
-        store.state.auth.reAuthenticationPromise == null ||
-        UtilitiesService.isObjectEmpty(store.state.auth.reAuthenticationPromise)
-      ) {
-        store.dispatch("auth/autoLogoutUser");
-        return Promise.reject("Auto Logout!");
-      } else {
-        return store.state.auth.reAuthenticationPromise;
-      }
-    }
+  async reAuthenticate(store) {
+    // set re Authentication state as in progress
+    store.dispatch("auth/setReAuthenticationState", "in-process");
 
-    // if the reAuthentication call is made while the user has been logged out
-    // then reject the promise
-    if (!store.getters["auth/isAuthenticated"])
-      return Promise.reject("User is not set!");
+    // create a promise and store it's resolver in a separate variable
+    let reAuthenticationPromiseResolver;
+    let reAuthenticationPromise = new Promise((resolve) => {
+      reAuthenticationPromiseResolver = resolve;
+    });
 
-    // set re-authentication as being in process
-    store.dispatch("auth/setReAuthenticationState", true);
+    // save the above created promise and it's resolver into the store
+    store.dispatch("auth/setReAuthenticationPromise", reAuthenticationPromise);
+    store.dispatch(
+      "auth/setReAuthenticationPromiseResolver",
+      reAuthenticationPromiseResolver
+    );
 
-    // start the refreshAccessToken call and save that promise
-    // in a variable called "reAuthenticationPromise"
-    const reAuthenticationPromise = UserAPIService.refreshAccessToken()
-      .then((response) => {
-        // once the refreshAccessToken call resolves, set the new access token in store
-        // set isReAuthenticating to false
-        // and unset the reAuthenticationPromise that was saved in store
-        // and resolve the promise so other waiting calls can proceed
+    try {
+      // try refreshing the access token
+      let response = await UserAPIService.refreshAccessToken();
+      if (response != undefined && response.data != undefined) {
+        // set the recieved access token into the store
+        // and mark the re authentication process as completed.
+        // We will also resolve the pending reAuthenticationPromise so the code
+        // that is waiting for the promises to resolve can proceed
         store.dispatch("auth/setAccessToken", response.data);
-        store.dispatch("auth/setReAuthenticationState", false);
-        store.dispatch("auth/unsetReAuthenticationPromise");
-        return Promise.resolve(true);
-      })
-      .catch((error) => {
-        return Promise.reject(error);
-      });
-
-    // Until the promise stored in "reAuthenticationPromise" is resolved,
-    // store this promise into vuex and return from the function.
-    store.dispatch("auth/setReAuthenticationPromise", reAuthenticationPromise);
-    return reAuthenticationPromise;
+        store.dispatch("auth/setReAuthenticationState", "completed");
```

Comment:
i don't see this state being used anywhere but maybe you wanted to log a state which was in-between in-progress, nor not-started? still, just think about its usage once.

Reply:
Yeah correct. It's not being used directly but if I HAD to write it, it would be kept here
```js
let reAuthenticationState = store.state.auth.reAuthenticationState;
          switch (reAuthenticationState) {
            case "dropped":
              // re-authentication process was dropped, throw an error.
              throw new Error(
                "Re-authentication failed, no more calls allowed"
              );
            case "in-process":
              // re authentication process in progress. Wait for the promise to resolve,
              // then proceed
              await store.state.auth.reAuthenticationPromise;
              break;
            case "not-started":
              // fresh re authentication process to begin now. Wait for the process to complete,
              // then proceed
              await UserFunctionalService.reAuthenticate(store);
              break;
            case "completed":
             break;
          }
```